Add spline support to ocean inputs

huwb opened the pull request:
**Status**: Ready to merge.

If Spline component attached alongside RegisterXYZInput componentflow generated along spline.

Put spline settings on spline. Can override on individual components.

Side fix: normal map flow animation had wrong offsets leading to weird animation

daleeidd commented:
> I think its more or less ok

It works well. I think many will be happy with it.

> except you'd really want flow speed control at each spline point, and i have not looked into how to have custom data on each spline point. Will probably take some work to do it in a generic way.

Hmmm. If the spline object stores all the information, that would work well I think. Then a custom editor for each spline point which proxies that data. Only downside is if the spline is deleted, then information is lost.

Otherwise, additional data objects similar to what Unity does. Simpler and safer.

> Oh dear i just realised this is branched off the dyn wave sim rebuild branch. I will have to separate this out later.. EDIT: Done now. It's also got some of the fix-me stuff that is posted in a separate PR which can be ignored.

I am also seeing commits for the fixed rate simulation.

huwb commented:
Been thinking about the custom data a bit more. Yeah centralising the data somewhere is a good thought... I don't want to allow any possibility that the spline points and custom data could become disconnected or out of sync though. The current setup where data is stored with each point means duplicating/deleting spline points works trivially and robustly with almost no code.

Maybe adding a new component for flow spline points which has custom data and is attached to spline points alongside the existing spline point component would work. Would be visible in inspector for editing, would support custom inspectors easily, and would also support handles etc easily.

huwb commented:
Oh the simulation changes should be reverted now (despite the commits being included). I'll probably squish this branch to clean it up later..

huwb commented:
I pushed an attempt at adding flow data to the spline in a generic way. i think i'm happy enough with this. Let me know if this looks ok to you.

It does not apply the speed properly if the spline is closed currently, can fix that later.

daleeidd commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterFlowInput.cs:
```diff
@@ -21,7 +27,20 @@ public class RegisterFlowInput : RegisterLodDataInputDisplacementCorrection<LodD
+        int _subdivisions = 1;
+        [SerializeField]
+        int _smoothingIterations = 0;
+        [SerializeField]
+        float _speed = 5f;
```
The behaviour of this property is uncertain with the name _speed. It appears that this behaves as max speed from what I tested (unless it is negative because of Mathf.Min). Not sure how you want to handle this. We could have a base speed and a separate property to set the max speed instead?

huwb replied:
It is intended to be the actual flow speed at that point. There is a bug however where if the spline is closed things are broken which may be confusing the matter. If your spline is open, this value should set the flow speed at hte spline point.

I should probably rename it to velocity as speed should not be negative, i will do that.

Lmk if this makes sense.

daleeidd replied:
Sorry. It doesn't.

> It is intended to be the actual flow speed at that point.

By this you mean _flowSpeed on flow spline point (not _speed on the spline)?

From the shader (_speed * _flowSpeed if I understand correctly)
```shader
return wt * _Speed * input.speed * input.axis;
```

Here _speed is just a multiplier. So if no flow data components are on the spline point then this will ultimately be zero and there will be no flow.

From the other issue:
> So if the component is not attached to a spline point, the default flow will be used.

It sounds like _speed is default flow here?

huwb replied:
Ok I think multiplying two speeds is making a mess here.

I'm going to remove the speed param from the material.

Then the only speed control will be the velocity values on the spline which will have a non zero default, as I just added in the latest commits.

So if you only create a vanilla spline it will have some fixed default velocity, then if you click the individual spline points you can change the default velocity.

huwb replied:
Done. There is now only one flow speed control - the Flow Velocity setting on the spline points, which has a reasonable default value. Let me know if this is clear now.

daleeidd replied:
Works well.

daleeidd commented on crest/Assets/Crest/Crest/Scripts/Spline/SplinePoint.cs:
```diff
@@ -63,6 +68,14 @@ public override void OnInspectorGUI()
                 return;
             }
 
+            if (parent.TryGetComponent(out ISplinePointCustomDataSetup customData))
+            {
+                if (customData.AttachDataToSplinePoint(thisSP.gameObject))
+                {
+                    EditorUtility.SetDirty(thisSP.gameObject);
+                }
+            }
```
Just a note. In the future we can use GetComponents instead to support more data types if that becomes a thing.

huwb replied:
Oh good point will do that

huwb replied:
Done, thanks

daleeidd reviewed (changes requested):
I've done a first pass review as I know it is unfinished. The approach is good. It looks like it could be easily extended if we ever get more data types that work with the spline too.

huwb commented on crest/Assets/Crest/Crest/Scripts/Spline/SplinePoint.cs:
```diff
@@ -63,6 +68,14 @@ public override void OnInspectorGUI()
                 return;
             }
 
+            if (parent.TryGetComponent(out ISplinePointCustomDataSetup customData))
+            {
+                if (customData.AttachDataToSplinePoint(thisSP.gameObject))
+                {
+                    EditorUtility.SetDirty(thisSP.gameObject);
+                }
+            }
```
Ah good point, i was going for something unorthodox here - I decided to make the code only weakly dependent on the data being present.

So if the component is not attached to a spline point, the default flow will be used. The mesh construction should handle this gracefully. If i want to change the flow somewhere, i click the spline point and the component attaches.

This is maybe not intuitive, or maybe a bit odd. Or it may fail in some weird ways. Let me know what you think.

daleeidd replied:
That makes sense. It's almost how Unity does it with the additional data components. So the issue really was that the default flow wasn't working? When I opened the scene there was no flow happening even though the flow input had a speed of 5.

huwb replied:
Oops there was indeed no default value when the data is not attached. It should be fixed now (defaults to 1, which is then multiplied by flow setting on material)

huwb commented:
I've added spline support to RegisterAnimWavesInput which is cool, can allow easy offsetting of water height, and also shows what is shared code that could be refactored out, or moved to a base class (pretty much all of it..).

The spline does not handle changes in height gracefully but it should be possible to make this work well.

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterLodDataInput.cs:
```diff
@@ -244,18 +247,106 @@ private void OnDrawGizmosSelected()
+
+                    if (_splineMaterial == null)
+                    {
+                        _splineMaterial = new Material(Shader.Find("Hidden/Crest/Inputs/Flow/Spline Geometry"));
```
Todo

daleeidd commented:
New stuff with the height is really cool.

huwb commented:
Ok i think its ready, could you please have a look @daleeidd 

I added a direct register height input as "anim waves input" is quite confusing for people i think.

daleeidd reviewed (changes requested):
Done a quick review. Just some small fixes for this time round. Works really well. Will do a proper test over the weekend.

One nice addition if it is not too complicated, is that when a splint point is selected, would it be possible to change the colour of the spline lines from the rest of the spline? Much easier to see what part of the spline is selected.

You can remove the test data if you're happy with it too. I'm happy with creating stuff from scratch when testing further.

huwb commented:
> when a splint point is selected, would it be possible to change the colour

good idea, done

> You can remove the test data

sounds good done

huwb commented on crest/Assets/Crest/Crest/Scripts/OceanValidation.cs:
```diff
@@ -67,35 +67,50 @@ public static void Suppressed(string message, MessageType type, Object @object =
-                );
+            var gameObject = lodInputComponent.targetObject as GameObject;
+            var renderer = gameObject.GetComponent<MeshRenderer>();
+            Object.DestroyImmediate(renderer);
```
didnt even know this was a thing. done!

daleeidd replied:
Same. I only just looked for it once I hit the bug. This might an issue elsewhere so I will do a pass in master.

huwb commented on crest/Assets/Crest/Crest/Scripts/OceanValidation.cs:
```diff
@@ -67,35 +67,50 @@ public static void Suppressed(string message, MessageType type, Object @object =
         {
             var gameObject = lodInputComponent.targetObject as GameObject;
-            gameObject.AddComponent<MeshRenderer>();
+            gameObject.AddComponent<Spline.Spline>();
```
didnt even know this was a thing too. done!

huwb commented on crest/Assets/Crest/Crest/Shaders/OceanInputs/Resources/FlowSplineGeometry.shader:
```diff
@@ -0,0 +1,100 @@
+﻿// Crest Ocean System
```
thanks done

huwb commented:
Merged with latest and did some associated fixup/cleanup.

Let me know what you think

daleeidd commented on crest/Assets/Crest/Crest/Scripts/LodData/SplinePointData/SplinePointDataNone.cs:
```diff
@@ -0,0 +1,21 @@
+    /// No data. This should not be attached to any apline point, but is used as a symbol
+    /// in the code when no data is required.
+    /// </summary>
+    public class SplinePointDataNone : MonoBehaviour, ISplinePointCustomData
```
Adding `[AddComponentMenu("")]` will hide it from the menu (I tried class visibility modifiers but they don't change anything). We should consider whether it is worthwhile adding it to the others too.

huwb replied:
interesting, did not know that. done. i guess it never needs to be added manually, if everything else is working

daleeidd commented on crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstnerBatched.cs:
```diff
@@ -853,7 +855,12 @@ public bool Validate(OceanRenderer ocean, ValidatedHelper.ShowMessage showMessag
             if (_mode == GerstnerMode.Geometry)
             {
-                isValid = ValidatedHelper.ValidateRenderer(gameObject, "Crest/Inputs/Animated Waves/Gerstner", showMessage);
+                isValid = ValidatedHelper.ValidateInputMesh(true, gameObject, showMessage);
+
+                if (TryGetComponent<MeshRenderer>(out var meshRenderer))
+                {
+                    isValid = ValidatedHelper.ValidateMaterial(meshRenderer.sharedMaterial, "Crest/Inputs/Animated Waves/Gerstner", gameObject, ValidatedHelper.DebugLog) && isValid;
+                }
```
Shouldn't this remain as _ValidateRenderer_ as it doesn't support splines?

huwb replied:
good point done

daleeidd reviewed (changes requested):
Nice. Just a few more things.

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterHeightInput.cs:
```diff
@@ -0,0 +1,71 @@
+
+        protected override string ShaderPrefix => "Crest/Inputs/Animated Waves";
+
+        protected override string SplineShaderName => "Crest/Inputs/Animated Waves/Add Water Height From Geometry";
```
thanks done

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterAnimWavesInput.cs:
```diff
@@ -12,10 +11,11 @@ namespace Crest
     [ExecuteAlways]
     [AddComponentMenu(MENU_PREFIX + "Animated Waves Input")]
-    public class RegisterAnimWavesInput : RegisterLodDataInput<LodDataMgrAnimWaves>
+    public class RegisterAnimWavesInput : RegisterLodDataInputWithSplineSupport<LodDataMgrAnimWaves>
```
I'm inclined to allow both to register scripts to work - i think that would be more intuitive for the user and i dont see a big downside - but let me know if you feel it should be disabled

daleeidd replied:
No worries. I cannot think of any real downsides either.

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterAnimWavesInput.cs:
```diff
@@ -31,7 +31,10 @@ public override float Wavelength
         protected override string ShaderPrefix => "Crest/Inputs/Animated Waves";
 
-        [SerializeField, Tooltip("Whether this input data should displace horizontally with waves. If false, data will not move from side to side with the waves. Adds a small performance overhead when disabled.")]
+        protected override string SplineShaderName => "Crest/Inputs/Animated Waves/Add Water Height From Geometry";
```
good catch done

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterHeightInput.cs:
```diff
@@ -0,0 +1,71 @@
+
+            maxDispVert = Mathf.Max(maxDispVert, _maxDisplacementVertical);
+
+            if (_maxDisplacementVertical > 0f)
```
that looks like a bug! it should be checking maxDispVert. fixed on both

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterHeightInput.cs:
```diff
@@ -0,0 +1,71 @@
+            // let ocean system know how far from the sea level this shape may displace the surface
+            if (_reportRendererBoundsToOceanSystem)
+            {
+                var minY = _renderer.bounds.min.y;
```
good idea - i added predication based on whether a component is attached, seems to work and disables most of the register input component fields when a spline is attached so i think it works well

daleeidd replied:
Nice. That works well.

huwb commented:
Thanks should be all addressed now

daleeidd reviewed (changes requested):
Nice. Almost there.

In addition to the minor change, the _#includes_ in _Set Base Water Height Using Geometry_ needs to be updated as it was moved to _Resources_. And _Set Water Height Using Geometry_ should be moved there too since it is also usds via scripting. We should leave a blank file in its original location just in case. And this could be an opportunity to rename the files too if you like since they are moved anyway?

huwb commented on crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstnerBatched.cs:
```diff
@@ -854,6 +856,11 @@ public bool Validate(OceanRenderer ocean, ValidatedHelper.ShowMessage showMessag
+                if (TryGetComponent<MeshRenderer>(out var meshRenderer))
+                {
+                    isValid = ValidatedHelper.ValidateMaterial(meshRenderer.sharedMaterial, "Crest/Inputs/Animated Waves/Gerstner", gameObject, ValidatedHelper.DebugLog) && isValid;
+                }
```
well caught, removed from here

huwb commented:
Good points

I added a 'stub' for the add height shader in the old location. It uses a meta file with the same name but different guid. i think this does the job, does it look right to you?

Fixed up the include path, cant believe this did not come up, even after reimporting..

I was not certain that "Set Water Height Using Geometry" is loaded from c#, could you confirm that this is the case?

daleeidd commented:
> I added a 'stub' for the add height shader in the old location, and renamed it in the new location. It uses a meta file with the same name but different guid. i think this does the job, does it look right to you?

Perfect. Looks correct.

> I was not certain that "Set Water Height Using Geometry" is loaded from c#, could you confirm that this is the case?

Sorry. Brain failure. You can ignore that one.

daleeidd reviewed (approved):
Looks great. Great effort too. I couldn't find any more issues. We already have some information on splines in wave conditions of the documentation. I'll just add a custom "preview" box for preview items and mention spline support is in preview. I'll add it to each appropriate simulation that has spline support and link to the general spline section. I can add it to this branch or master after merging?

huwb commented:
Yeah i thought about docs and decided to focus on getting the code merged for now, then step back and review docs later. I think its a bit easier for my brain to tackle things in separate stages.

The preview box sounds like a good idea. Or maybe its ready to release it? It might also be that we just un-preview it for the next release.

I'll see if i can get someone testing it now that its merged which will give us more data/confidence as well.

daleeidd commented:
No worries. Possibly. Will discuss on Discord.
